File: searxng_search.mjs
#!/usr/bin/env node

/**
 * SearXNG Backend for Hermes Agent — Token-optimized
 * Recherche web + extraction via instance SearXNG auto-hébergée.
 */

import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const https = require('https');
const http = require('http');
const { URL } = require('url');

const SEARXNG_URL = (process.env.SEARXNG_URL || '').replace(/\/+$/, '');
const SEARXNG_LANGUAGE = process.env.SEARXNG_LANGUAGE || 'fr';
const SEARXNG_SAFESARCH = parseInt(process.env.SEARXNG_SAFESARCH || '1');
const SEARXNG_MAX_RESULTS = parseInt(process.env.SEARXNG_MAX_RESULTS || '5');
const SEARXNG_EXTRACT_MAX = parseInt(process.env.SEARXNG_EXTRACT_MAX || '5000');
const SEARXNG_TIMEOUT = parseInt(process.env.SEARXNG_TIMEOUT || '15') * 1000;

class SearXNGClient {
  constructor(url) {
    this.url = url || SEARXNG_URL;
    if (!this.url) throw new Error('SEARXNG_URL non configuré');
  }

  async search(query, options = {}) {
    const {
      limit = SEARXNG_MAX_RESULTS,
      language = SEARXNG_LANGUAGE,
      timeRange = null,
      categories = 'general',
    } = options;

    const url = new URL(`${this.url}/search`);
    url.searchParams.append('q', query);
    url.searchParams.append('format', 'json');
    url.searchParams.append('language', language);
    url.searchParams.append('safesearch', SEARXNG_SAFESARCH);
    url.searchParams.append('categories', categories);
    if (timeRange) url.searchParams.append('time_range', timeRange);

    const data = await this._fetch(url.toString());
    const web = (data.results || []).slice(0, limit).map((r, i) => ({
      title: (r.title || '').trim(),
      url: r.url || '',
      description: (r.content || '').trim().substring(0, 200),
      position: i + 1,
    })).filter(r => r.url);

    return { success: true, data: { web } };
  }

  async extract(urls, maxLen = SEARXNG_EXTRACT_MAX) {
    if (typeof urls === 'string') urls = [urls];
    const pages = await Promise.all(urls.map(async (targetUrl) => {
      try {
        const html = await this._fetchRaw(targetUrl);
        const titleMatch = html.match(/<title[^>]*>(.*?)<\/title>/is);
        const title = titleMatch ? titleMatch[1].trim() : targetUrl;

        let text = html
          .replace(/<(script|style|nav|footer|header|aside|noscript)[^>]*>[\s\S]*?<\/\1>/gi, '')
          .replace(/<[^>]+>/g, ' ')
          .replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>')
          .replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&nbsp;/g, ' ')
          .replace(/\s+/g, ' ').trim();

        if (text.length > maxLen) {
          const half = Math.floor(maxLen / 2);
          text = text.substring(0, half) + '\n\n[...truncated...]\n\n' + text.substring(text.length - half);
        }

        return { url: targetUrl, title, content: text, success: true };
      } catch (e) {
        return { url: targetUrl, error: e.message.substring(0, 200), success: false };
      }
    }));

    return { success: pages.some(p => p.success), pages };
  }

  _fetch(url) {
    return new Promise((resolve, reject) => {
      const proto = url.startsWith('https') ? https : http;
      const req = proto.get(url, { headers: { 'User-Agent': 'Hermes-SearXNG/1.0', 'Accept': 'application/json' } }, (res) => {
        let data = '';
        res.on('data', c => data += c);
        res.on('end', () => { try { resolve(JSON.parse(data)); } catch (e) { reject(new Error('Invalid JSON')); } });
      });
      req.on('error', reject);
      req.setTimeout(SEARXNG_TIMEOUT, () => { req.destroy(); reject(new Error('Timeout')); });
    });
  }

  _fetchRaw(url) {
    return new Promise((resolve, reject) => {
      const proto = url.startsWith('https') ? https : http;
      const req = proto.get(url, { headers: { 'User-Agent': 'Mozilla/5.0 (compatible; Hermes-SearXNG/1.0)' } }, (res) => {
        let data = '';
        res.on('data', c => data += c);
        res.on('end', () => resolve(data));
      });
      req.on('error', reject);
      req.setTimeout(SEARXNG_TIMEOUT, () => { req.destroy(); reject(new Error('Timeout')); });
    });
  }

  formatSearch(results) {
    const web = results?.data?.web;
    if (!web?.length) return 'Aucun résultat trouvé.';
    return `\n🔍 ${web.length} résultat(s)\n\n` +
      web.map(r => `${r.position}. **${r.title}**\n   ${r.url}\n   ${r.description || ''}\n`).join('\n');
  }

  formatExtract(result) {
    return (result.pages || []).map(p =>
      p.success ? `# ${p.title}\n\n${p.content}\n\n---\n` : `❌ ${p.url}: ${p.error}\n---\n`
    ).join('\n');
  }
}

// CLI
async function main() {
  const args = process.argv.slice(2);
  const extractIdx = args.indexOf('--extract');
  const jsonIdx = args.indexOf('--json');

  if (extractIdx !== -1) {
    const urls = args.slice(extractIdx + 1).filter(a => !a.startsWith('--'));
    const client = new SearXNGClient();
    const result = await client.extract(urls);
    console.log(jsonIdx !== -1 ? JSON.stringify(result, null, 2) : client.formatExtract(result));
  } else {
    const query = args.filter(a => !a.startsWith('--')).join(' ');
    if (!query) {
      console.error('Usage: node searxng_search.mjs "recherche" [--json] [--extract URL...]');
      process.exit(1);
    }
    const client = new SearXNGClient();
    const result = await client.search(query);
    console.log(jsonIdx !== -1 ? JSON.stringify(result, null, 2) : client.formatSearch(result));
  }
}

export { SearXNGClient };

if (import.meta.url === `file://${process.argv[1]}`) main();

File: searxng_search.py
#!/usr/bin/env python3
"""
SearXNG Backend for Hermes Agent — Optimisé tokens
Recherche web + extraction via instance SearXNG auto-hébergée.
Zéro API key, stdlib uniquement, outputs compacts.

Formats de recherche:
  compact  — titre + URL + description courte (défaut, ~327 tokens/5 résultats)
  rich     — + date, source, score, thumbnail (~550 tokens/5 résultats)
"""

import os
import sys
import json
import re
import urllib.request
import urllib.parse
import html as html_mod
from urllib.error import URLError, HTTPError

# ─── Configuration ────────────────────────────────────────────────────────────
SEARXNG_URL = os.environ.get("SEARXNG_URL", "").rstrip("/")
SEARXNG_TIMEOUT = int(os.environ.get("SEARXNG_TIMEOUT", "15"))
SEARXNG_LANGUAGE = os.environ.get("SEARXNG_LANGUAGE", "fr")
SEARXNG_SAFESEARCH = int(os.environ.get("SEARXNG_SAFESEARCH", "1"))
SEARXNG_MAX_RESULTS = int(os.environ.get("SEARXNG_MAX_RESULTS", "5"))
SEARXNG_DESC_MAX = int(os.environ.get("SEARXNG_DESC_MAX", "200"))
SEARXNG_EXTRACT_MAX = int(os.environ.get("SEARXNG_EXTRACT_MAX", "5000"))
SEARXNG_EXTRACT_TIMEOUT = int(os.environ.get("SEARXNG_EXTRACT_TIMEOUT", "20"))


class SearXNGClient:
    """Client SearXNG optimisé pour Hermes Agent — outputs compacts, minimum tokens."""

    def __init__(self, url=None, language=None, timeout=None):
        self.url = url or SEARXNG_URL
        self.language = language or SEARXNG_LANGUAGE
        self.timeout = timeout or SEARXNG_TIMEOUT
        if not self.url:
            raise ValueError(
                "SEARXNG_URL non configuré. "
                "Définissez SEARXNG_URL ou passez url= au constructeur."
            )

    # ─── Recherche ────────────────────────────────────────────────────────────
    def search(self, query, limit=None, language=None, time_range=None,
               categories="general", fmt="compact"):
        """
        Recherche SearXNG — retourne un dict optimisé pour LLM.

        Formats:
          compact — titre, url, description tronquée (défaut)
          rich    — + publishedDate, engine, score, thumbnail

        Format de retour (compatible Hermes web_search_tool) :
        {
            "success": True,
            "data": {
                "web": [
                    {"title": "...", "url": "...", "description": "...", "position": 1,
                     "publishedDate": "...", "engine": "...", "score": 6.8, "thumbnail": "..."},
                    ...
                ]
            }
        }
        """
        limit = limit or SEARXNG_MAX_RESULTS
        language = language or self.language

        params = {
            "q": query,
            "format": "json",
            "language": language,
            "safesearch": SEARXNG_SAFESEARCH,
            "categories": categories,
        }
        if time_range:
            params["time_range"] = time_range

        url = f"{self.url}/search?{urllib.parse.urlencode(params)}"
        req = urllib.request.Request(url)
        req.add_header("User-Agent", "Hermes-SearXNG/1.0")
        req.add_header("Accept", "application/json")

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except HTTPError as e:
            raise Exception(f"SearXNG HTTP {e.code}: {e.reason}")
        except URLError as e:
            raise Exception(f"SearXNG connexion: {e.reason}")
        except json.JSONDecodeError:
            raise Exception("SearXNG: réponse JSON invalide")

        web_results = []
        for i, r in enumerate(data.get("results", [])[:limit]):
            desc = html_mod.unescape(r.get("content", "").strip())
            if len(desc) > SEARXNG_DESC_MAX:
                desc = desc[:SEARXNG_DESC_MAX] + "…"

            # Base entry (compact)
            entry = {
                "title": r.get("title", "").strip(),
                "url": r.get("url", ""),
                "description": desc,
                "position": i + 1,
            }

            # Rich format: ajouter métadonnées
            if fmt == "rich":
                if r.get("publishedDate"):
                    entry["publishedDate"] = r["publishedDate"]
                if r.get("engine"):
                    entry["engine"] = r["engine"]
                if r.get("engines"):
                    entry["engines"] = r["engines"]
                if r.get("score"):
                    entry["score"] = r["score"]
                if r.get("thumbnail"):
                    entry["thumbnail"] = r["thumbnail"]

            if entry["url"]:
                web_results.append(entry)

        return {"success": True, "data": {"web": web_results}}

    # ─── Extraction ───────────────────────────────────────────────────────────
    def extract(self, urls, max_length=None):
        """
        Extraction de contenu web — HTML nettoyé en texte compact.

        Pas de dépendance externe (no beautifulsoup, no readability, no firecrawl).
        Utilise urllib + regex pour un nettoyage basique mais efficace.

        Format de retour (compatible Hermes web_extract_tool) :
        {
            "success": True,
            "pages": [
                {"url": "...", "title": "...", "content": "...", "success": True},
                ...
            ]
        }
        """
        max_length = max_length or SEARXNG_EXTRACT_MAX
        if isinstance(urls, str):
            urls = [urls]

        pages = []
        for target_url in urls:
            try:
                req = urllib.request.Request(target_url)
                req.add_header("User-Agent",
                    "Mozilla/5.0 (compatible; Hermes-SearXNG/1.0)")
                with urllib.request.urlopen(req, timeout=SEARXNG_EXTRACT_TIMEOUT) as resp:
                    raw = resp.read()
                    # Try multiple encodings
                    for enc in ("utf-8", "latin-1", "cp1252"):
                        try:
                            html_content = raw.decode(enc)
                            break
                        except (UnicodeDecodeError, ValueError):
                            continue
                    else:
                        html_content = raw.decode("utf-8", errors="replace")

                # Extraction du titre
                title_match = re.search(
                    r"<title[^>]*>(.*?)</title>", html_content,
                    re.DOTALL | re.IGNORECASE
                )
                title = html_mod.unescape(title_match.group(1).strip()) if title_match else target_url

                # Nettoyage HTML → texte
                text = html_content
                for tag in ("script", "style", "nav", "footer", "header", "aside", "noscript"):
                    text = re.sub(
                        rf"<{tag}[^>]*>.*?</{tag}>", "", text,
                        flags=re.DOTALL | re.IGNORECASE
                    )
                text = re.sub(r"<[^>]+>", " ", text)
                text = html_mod.unescape(text)
                text = re.sub(r"\s+", " ", text).strip()

                # Troncature intelligente
                if len(text) > max_length:
                    half = max_length // 2
                    text = text[:half] + "\n\n[...contenu tronqué...]\n\n" + text[-half:]

                pages.append({
                    "url": target_url,
                    "title": title,
                    "content": text,
                    "success": True,
                })
            except Exception as e:
                pages.append({
                    "url": target_url,
                    "error": str(e)[:200],
                    "success": False,
                })

        return {"success": any(p.get("success") for p in pages), "pages": pages}

    # ─── Format CLI ───────────────────────────────────────────────────────────
    def format_search(self, results, fmt="compact"):
        """Format lisible pour le CLI."""
        if not results.get("data", {}).get("web"):
            return "Aucun résultat trouvé."

        web = results["data"]["web"]
        lines = [f"🔍 {len(web)} résultat(s)\n"]

        if fmt == "rich":
            for r in web:
                lines.append(f"{r['position']}. **{r['title']}**")
                lines.append(f"   🔗 {r['url']}")
                if r.get("description"):
                    lines.append(f"   📄 {r['description']}")
                if r.get("publishedDate"):
                    lines.append(f"   📅 {r['publishedDate']}")
                if r.get("engine"):
                    engines = r.get("engines", [r["engine"]])
                    lines.append(f"   🔍 {' | '.join(engines)}")
                if r.get("score"):
                    lines.append(f"   ⭐ {r['score']}")
                if r.get("thumbnail"):
                    lines.append(f"   🖼️ {r['thumbnail'][:80]}…")
                lines.append("")
        else:
            for r in web:
                lines.append(f"{r['position']}. [{r['title']}]({r['url']})")
                if r.get("description"):
                    lines.append(f"   {r['description']}")
                lines.append("")

        return "\n".join(lines)

    def format_extract(self, result):
        """Format lisible pour extraction CLI."""
        lines = []
        for page in result.get("pages", []):
            if page.get("success"):
                lines.append(f"# {page['title']}\n")
                lines.append(page["content"])
            else:
                lines.append(f"❌ {page['url']}: {page.get('error', 'Unknown')}")
            lines.append("\n---\n")
        return "\n".join(lines)


# ─── CLI ──────────────────────────────────────────────────────────────────────
def main():
    import argparse

    parser = argparse.ArgumentParser(
        description="SearXNG Backend pour Hermes Agent — Optimisé tokens"
    )
    parser.add_argument("query", nargs="?", help="Requête de recherche")
    parser.add_argument("--extract", nargs="+", metavar="URL",
                        help="Extraire le contenu d'URL(s)")
    parser.add_argument("--format", choices=["compact", "rich"], default="compact",
                        help="Format de sortie (compact=défaut, rich=métadonnées)")
    parser.add_argument("--language", default="fr", help="Code langue (défaut: fr)")
    parser.add_argument("--time-range", choices=["day", "week", "month", "year"],
                        help="Filtrer par période")
    parser.add_argument("--limit", type=int, default=5, help="Nombre max de résultats")
    parser.add_argument("--categories", default="general",
                        help="Catégories: general, images, videos, etc.")
    parser.add_argument("--json", action="store_true",
                        help="Sortie JSON brute (pour intégration Hermes)")
    parser.add_argument("--url", help="URL instance SearXNG (override SEARXNG_URL)")

    args = parser.parse_args()

    if not args.query and not args.extract:
        parser.error("Spécifiez une requête ou --extract URL")

    client = SearXNGClient(url=args.url)

    if args.extract:
        result = client.extract(args.extract)
        if args.json:
            print(json.dumps(result, indent=2, ensure_ascii=False))
        else:
            print(client.format_extract(result))
    else:
        result = client.search(
            args.query,
            limit=args.limit,
            language=args.language,
            time_range=args.time_range,
            categories=args.categories,
            fmt=args.format,
        )
        if args.json:
            print(json.dumps(result, indent=2, ensure_ascii=False))
        else:
            print(client.format_search(result, fmt=args.format))


if __name__ == "__main__":
    main()